File: app/src/main/java/dev/danielholmberg/improve/Improve.java
package dev.danielholmberg.improve;

import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.os.Build;
import android.os.Environment;
import android.util.Log;

import androidx.fragment.app.Fragment;

import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.remoteconfig.FirebaseRemoteConfig;
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import dev.danielholmberg.improve.Activities.MainActivity;
import dev.danielholmberg.improve.Adapters.ArchivedNotesAdapter;
import dev.danielholmberg.improve.Adapters.CompanyRecyclerViewAdapter;
import dev.danielholmberg.improve.Adapters.ContactRecyclerViewAdapter;
import dev.danielholmberg.improve.Adapters.NotesAdapter;
import dev.danielholmberg.improve.Adapters.TagsAdapter;
import dev.danielholmberg.improve.Adapters.VipImagesAdapter;
import dev.danielholmberg.improve.Fragments.ArchivedNotesFragment;
import dev.danielholmberg.improve.Fragments.ContactsFragment;
import dev.danielholmberg.improve.Fragments.NoteDetailsDialogFragment;
import dev.danielholmberg.improve.Fragments.NotesFragment;
import dev.danielholmberg.improve.Managers.AuthManager;
import dev.danielholmberg.improve.Managers.FirebaseDatabaseManager;
import dev.danielholmberg.improve.Managers.FirebaseStorageManager;
import dev.danielholmberg.improve.Services.DriveServiceHelper;

/**
 * Created by Daniel Holmberg.
 */

public class Improve extends Application implements Serializable {

    private AuthManager authManager;
    private FirebaseDatabaseManager firebaseDatabaseManager;
    private FirebaseStorageManager firebaseStorageManager;
    private FirebaseRemoteConfig firebaseRemoteConfig;
    private DriveServiceHelper mDriveServiceHelper;

    // volatile attribute makes the singleton thread safe.
    private static volatile Improve sImproveInstance;

    private File rootDir, imageDir;

    public MainActivity mainActivityRef;

    private NotesFragment notesFragmentRef;
    private ContactsFragment contactsFragmentRef;
    private ArchivedNotesFragment archivedNotesFragmentRef;

    private NotesAdapter notesAdapter;
    private ArchivedNotesAdapter archivedNotesAdapter;
    private TagsAdapter tagsAdapter;
    private Fragment currentFragment;
    private CompanyRecyclerViewAdapter companyRecyclerViewAdapter;
    private final HashMap<String, ContactRecyclerViewAdapter> contactAdapters = new HashMap<>();

    private boolean isVIPUser = false;
    public final Map VIP_USERS = new HashMap<String, Object>() {{
        put("1", "danielholmberg.dev@gmail.com");
        put("2", "danielkurtholmberg@gmail.com");
    }};

    public static Improve getInstance() {
        // Double checks locking to prevent unnecessary sync.
        if (sImproveInstance == null) {
            synchronized (Improve.class) {
                // If there is no instance available... create new one
                if (sImproveInstance == null) {
                    sImproveInstance = new Improve();
                }
            }
        }
        return sImproveInstance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        sImproveInstance = this;
        // Enabling offline capabilities for Firebase Storage.
        // OBS!!! Can create Local Firebase cache issue where data changed in console won't take effect.
        // Reset cache by setting this to FASLE if there is an issue with data out of sync,
        // or a crash due to changed Model parameters.
        FirebaseDatabase.getInstance().setPersistenceEnabled(true);

        // Create Root-Dir if not already existing.
        getRootDir();

        // Initializing managers.
        authManager = new AuthManager();
        firebaseDatabaseManager = new FirebaseDatabaseManager();
        firebaseStorageManager = new FirebaseStorageManager();

        firebaseRemoteConfig = FirebaseRemoteConfig.getInstance();
        FirebaseRemoteConfigSettings configSettings = new FirebaseRemoteConfigSettings.Builder()
                .setMinimumFetchIntervalInSeconds(BuildConfig.DEBUG ? 0 : 3600)
                .build();
        firebaseRemoteConfig.setConfigSettingsAsync(configSettings);
        firebaseRemoteConfig.setDefaults(R.xml.remote_config_defaults);
        firebaseRemoteConfig.setDefaults(VIP_USERS);

        createNotificationChannelExport();
    }

    public void setIsVIPUser(boolean isVIPUser) {
        this.isVIPUser = isVIPUser;
    }

    public boolean isVIPUser() {
        return this.isVIPUser;
    }

    private void createNotificationChannelExport() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            String id = getString(R.string.export_channel_id);
            CharSequence name = getString(R.string.export_channel_name);
            String description = getString(R.string.export_channel_desc);
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            NotificationChannel channel = new NotificationChannel(id, name, importance);
            channel.setDescription(description);
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            if (notificationManager != null) {
                notificationManager.createNotificationChannel(channel);
            }
        }
    }

    /**
     * Returns the Firebase authentication manager.
     * @return - Authentication manager of Firebase
     */
    public AuthManager getAuthManager() {
        return this.authManager;
    }

    /**
     * Returns the application specific Firebase database manager.
     * @return - Database manager of Firebase
     */
    public FirebaseDatabaseManager getFirebaseDatabaseManager() {
        return this.firebaseDatabaseManager;
    }

    /**
     * Returns the application specific Firebase Cloud Storage manager.
     * @return - Cloud Storage manager of Firebase
     */
    public FirebaseStorageManager getFirebaseStorageManager() {
        return this.firebaseStorageManager;
    }

    /**
     * Returns the application specific Firebase remote config.
     * @return - FirebaseRemoteConfig singleton object.
     */
    public FirebaseRemoteConfig getFirebaseRemoteConfig() {
        return this.firebaseRemoteConfig;
    }

    public void saveState() {
        if(getTagsAdapter() != null) {
            getFirebaseDatabaseManager().saveTags(getTags());
        }
        if(getNotesAdapter() != null) {
            getFirebaseDatabaseManager().saveNotes(getNotes());
        }
        if(getArchivedNotesAdapter() != null) {
            getFirebaseDatabaseManager().saveArchivedNotes(getArchivedNotes());
        }
        if(getCompanyRecyclerViewAdapter() != null) {
            getFirebaseDatabaseManager().saveCompanies(getCompanies());
        }
    }

    /**
     * Returns the application specific Root directory.
     * @return - Root directory of application
     */
    public File getRootDir() {
        rootDir = getApplicationContext().getFilesDir();
        if(!rootDir.exists()) {
            rootDir.mkdirs();
        }
        Log.d("Improve", "RootDir: " + this.rootDir.getPath());
        return this.rootDir;
    }

    public File getImageDir() {
        imageDir = new File(rootDir, FirebaseStorageManager.IMAGES_REF);
        if(!imageDir.exists()) {
            imageDir.mkdirs();
        }
        Log.d("Improve", "ImageDir: " + this.imageDir.getPath());
        return this.imageDir;
    }

    public MainActivity getMainActivityRef() {
        return mainActivityRef;
    }

    public void setMainActivityRef(MainActivity mainActivity) {
        this.mainActivityRef = mainActivity;
    }

    public Fragment getCurrentFragment() {
        return currentFragment;
    }

    public void setCurrentFragment(Fragment currentFragment) {
        this.currentFragment = currentFragment;
    }

    // ---- Note functions ---- //

    public void setNotesFragmentRef(NotesFragment notesFragmentRef) {
        this.notesFragmentRef = notesFragmentRef;
    }

    public NotesFragment getNotesFragmentRef() {
        return this.notesFragmentRef;
    }

    public void setNotesAdapter(NotesAdapter notesAdapter) {
        this.notesAdapter = notesAdapter;
    }

    public NotesAdapter getNotesAdapter() {
        return notesAdapter;
    }

    public HashMap<String, Object> getNotes() {
        return getNotesAdapter().getHashMap();
    }

    // ---- Archived note functions ---- //

    public void setArchivedNotesFragmentRef(ArchivedNotesFragment archivedNotesFragmentRef) {
        this.archivedNotesFragmentRef = archivedNotesFragmentRef;
    }

    public ArchivedNotesFragment getArchivedNotesFragmentRef() {
        return this.archivedNotesFragmentRef;
    }

    public void setArchivedNotesAdapter(ArchivedNotesAdapter archivedNotesAdapter) {
        this.archivedNotesAdapter = archivedNotesAdapter;
    }

    public ArchivedNotesAdapter getArchivedNotesAdapter() {
        return archivedNotesAdapter;
    }

    public HashMap<String,Object> getArchivedNotes() {
        return getArchivedNotesAdapter().getHashMap();
    }

    // ---- Tag functions ---- //

    public void setTagsAdapter(TagsAdapter tagsAdapter) {
        this.tagsAdapter = tagsAdapter;
    }

    public TagsAdapter getTagsAdapter() {
        return tagsAdapter;
    }

    public HashMap<String,Object> getTags() {
        return getTagsAdapter().getHashMap();
    }

    // ---- Contact functions ---- //

    public void setContactFragmentRef(ContactsFragment contactFragmentRef) {
        this.contactsFragmentRef = contactFragmentRef;
    }

    public ContactsFragment getContactsFragmentRef() {
        return this.contactsFragmentRef;
    }

    // ---- Company functions ---- //

    public void setCompanyRecyclerViewAdapter(CompanyRecyclerViewAdapter companyRecyclerViewAdapter) {
        this.companyRecyclerViewAdapter = companyRecyclerViewAdapter;
    }

    public CompanyRecyclerViewAdapter getCompanyRecyclerViewAdapter() {
        return companyRecyclerViewAdapter;
    }

    public HashMap<String, Object> getCompanies() {
        return getCompanyRecyclerViewAdapter().getCompaniesHashMap();
    }

    public void addContactsAdapter(String nameId, ContactRecyclerViewAdapter contactsAdapter) {
        contactAdapters.put(nameId, contactsAdapter);
    }

    public ContactRecyclerViewAdapter getCompanyContactsAdapter(String companyId) {
        return contactAdapters.get(companyId);
    }

    public void setDriveServiceHelper(DriveServiceHelper mDriveServiceHelper) {
        this.mDriveServiceHelper = mDriveServiceHelper;
    }

    public DriveServiceHelper getDriveServiceHelper() {
        return mDriveServiceHelper;
    }
}

File: app/src/main/java/dev/danielholmberg/improve/Improve.kt
package dev.danielholmberg.improve

import android.app.Application
import dev.danielholmberg.improve.Managers.AuthManager
import dev.danielholmberg.improve.Managers.DatabaseManager
import dev.danielholmberg.improve.Managers.StorageManager
import dev.danielholmberg.improve.Services.DriveServiceHelper
import dev.danielholmberg.improve.Activities.MainActivity
import dev.danielholmberg.improve.Fragments.NotesFragment
import dev.danielholmberg.improve.Fragments.ContactsFragment
import dev.danielholmberg.improve.Fragments.ArchivedNotesFragment
import dev.danielholmberg.improve.Adapters.NotesAdapter
import dev.danielholmberg.improve.Adapters.ArchivedNotesAdapter
import dev.danielholmberg.improve.Adapters.TagsAdapter
import dev.danielholmberg.improve.Adapters.CompanyRecyclerViewAdapter
import dev.danielholmberg.improve.Adapters.ContactRecyclerViewAdapter
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.remoteconfig.FirebaseRemoteConfigSettings
import android.os.Build
import android.app.NotificationManager
import android.app.NotificationChannel
import android.util.Log
import androidx.fragment.app.Fragment
import com.google.firebase.remoteconfig.FirebaseRemoteConfig
import dev.danielholmberg.improve.Managers.RemoteConfigManager
import dev.danielholmberg.improve.Services.SharedPrefsService
import java.io.File
import java.io.Serializable
import java.util.*
import javax.inject.Singleton
import kotlin.jvm.Volatile

@Singleton
class Improve : Application(), Serializable {

    lateinit var authManager: AuthManager
        private set
    lateinit var databaseManager: DatabaseManager
        private set
    lateinit var storageManager: StorageManager
        private set
    lateinit var remoteConfigManager: RemoteConfigManager
        private set

    var driveServiceHelper: DriveServiceHelper? = null

    var mainActivityRef: MainActivity? = null

    lateinit var imageDir: File
        private set

    var notesFragmentRef: NotesFragment? = null
    var contactsFragmentRef: ContactsFragment? = null

    var archivedNotesFragmentRef: ArchivedNotesFragment? = null
    var notesAdapter: NotesAdapter? = null
    var archivedNotesAdapter: ArchivedNotesAdapter? = null

    var tagsAdapter: TagsAdapter? = null
    var currentFragment: Fragment? = null

    var companyRecyclerViewAdapter: CompanyRecyclerViewAdapter? = null
    private val contactAdapters = HashMap<String, ContactRecyclerViewAdapter>()

    val notes: HashMap<String?, Any>
        get() = notesAdapter!!.hashMap
    val archivedNotes: HashMap<String?, Any>
        get() = archivedNotesAdapter!!.hashMap
    val tags: HashMap<String?, Any>
        get() = tagsAdapter!!.hashMap
    val companies: HashMap<String, Any>
        get() = companyRecyclerViewAdapter!!.companiesHashMap

    @set:JvmName("setVipUser")
    var isVipUser: Boolean = false

    private val vipUsers: Map<String, String> = object : HashMap<String, String>() {
        init {
            put("1", "danielholmberg.dev@gmail.com")
            put("2", "danielkurtholmberg@gmail.com")
        }
    }

    private lateinit var sharedPrefsService: SharedPrefsService
    lateinit var deviceId: String

    override fun onCreate() {
        super.onCreate()
        sImproveInstance = this

        loadPreferences()
        enableFirebaseStorageOfflineCapabilities()

        val rootDir = generateRootDir()
        generateImageDir(rootDir)

        authManager = AuthManager()
        databaseManager = DatabaseManager()
        storageManager = StorageManager()
        remoteConfigManager = RemoteConfigManager(remoteConfig = FirebaseRemoteConfig.getInstance())

        setRemoteConfigSettings()

        createNotificationChannelExport()
    }

    private fun enableFirebaseStorageOfflineCapabilities() {
        // Enabling offline capabilities for Firebase Storage.
        // OBS!!! Can create Local Firebase cache issue where data changed in console won't take effect.
        // Reset cache by setting this to FALSE if there is an issue with data out of sync,
        // or a crash due to changed Model parameters.
        FirebaseDatabase.getInstance().setPersistenceEnabled(true)
    }

    private fun loadPreferences() {
        sharedPrefsService = SharedPrefsService(this, "DEVICE_PREFS")

        // Retrieve the stored Device ID or generate a new
        val deviceIdKey = "device_id"
        val storedDeviceId = sharedPrefsService.getString(deviceIdKey)
        if ((storedDeviceId == null) || storedDeviceId.isEmpty()) {
            deviceId = UUID.randomUUID().toString()
            sharedPrefsService.putString(deviceIdKey, deviceId)
        } else {
            deviceId = storedDeviceId
        }

        Log.i(TAG, "Device ID: $deviceId")
    }

    private fun setRemoteConfigSettings() {
        val configSettings = FirebaseRemoteConfigSettings.Builder()
            .setMinimumFetchIntervalInSeconds(if (BuildConfig.DEBUG) 0 else 3600.toLong())
            .build()
        remoteConfigManager.setConfigSettingsAsync(configSettings)
        remoteConfigManager.setDefaultsAsync(R.xml.remote_config_defaults)
        remoteConfigManager.setDefaultsAsync(vipUsers)
    }

    private fun createNotificationChannelExport() {
        // Create the NotificationChannel, but only on API 26+ because
        // the NotificationChannel class is new and not in the support library
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val id = getString(R.string.export_channel_id)
            val name: CharSequence = getString(R.string.export_channel_name)
            val description = getString(R.string.export_channel_desc)
            val importance = NotificationManager.IMPORTANCE_DEFAULT
            val channel = NotificationChannel(id, name, importance)
            channel.description = description
            // Register the channel with the system; you can't change the importance
            // or other notification behaviors after this
            val notificationManager = getSystemService(
                NotificationManager::class.java
            )
            notificationManager?.createNotificationChannel(channel)
        }
    }

    fun saveState() {
        tagsAdapter?.let { databaseManager.saveTags(tags) }
        notesAdapter?.let { databaseManager.saveNotes(notes) }
        archivedNotesAdapter?.let { databaseManager.saveArchivedNotes(archivedNotes) }
        companyRecyclerViewAdapter?.let { databaseManager.saveCompanies(companies) }
    }

    /**
     * Returns the application specific Root directory.
     * @return - Root directory of application
     */
    private fun generateRootDir(): File {
        val rootDir = applicationContext.filesDir
        if (!rootDir.exists()) {
            rootDir.mkdirs()
        }
        Log.d("Improve", "RootDir: " + rootDir.path)
        return rootDir
    }

    private fun generateImageDir(rootDir: File): File {
        imageDir = File(rootDir, StorageManager.IMAGES_REF)
        if (!imageDir.exists()) {
            imageDir.mkdirs()
        }
        Log.d("Improve", "ImageDir: " + imageDir.path)
        return imageDir
    }

    fun addContactsAdapter(nameId: String, contactsAdapter: ContactRecyclerViewAdapter) {
        contactAdapters[nameId] = contactsAdapter
    }

    fun getCompanyContactsAdapter(companyId: String): ContactRecyclerViewAdapter? {
        return contactAdapters[companyId]
    }

    companion object {
        private val TAG = this::class.simpleName

        // volatile attribute makes the singleton thread safe.
        @Volatile
        private var sImproveInstance: Improve? = null

        @JvmStatic
        val instance: Improve?
            get() {
                // Double checks locking to prevent unnecessary sync.
                if (sImproveInstance == null) {
                    synchronized(Improve::class.java) {
                        // If there is no instance available... create new one
                        if (sImproveInstance == null) {
                            sImproveInstance = Improve()
                        }
                    }
                }
                return sImproveInstance
            }
    }
}